### ISSUE TITLE
Training a NODE using multiple shooting. 

### ISSUE THREAD

**Issue by ashaygan** [SciML/DiffEqFlux.jl#717]
I am trying to fit an NODE to the data given in the following example:
https://diffeqflux.sciml.ai/stable/examples/multiple_shooting/
except that the task is to learn only one time series, i.e. the first dimension of the two dimensional data produced by this function:
```
function trueODEfunc(du, u, p, t)
    true_A = [-0.1 2.0; -2.0 -0.1]
    du .= ((u.^3)'true_A)'
end 
``` 
It seems the task is easier now since the network needs to fit to one dimension less. However, it fails to fit to the data no matter how large the network is (I didn't go too far! 7, 8 layers, 100n per layer, and some activation functions tested). The problem is actually harder because of two reasons I can imagine: 

1. By learning two dimensions, we lift the space into higher than one dimension and it may help the process of learning one dimension separately (regardless of whether or not we learn the other one). Kind of the same as using an augmented NODE. 
2. The data in the second dimension may actually help us and lead us to learn the second dimension, leaving aside the positive effect of moving to the higher dimension. But also it may mislead us and makes the learning process hard!!!

Now I'd like to know if the above reasoning in failing a network to only one of the dimensions is correct, if there are more reasons than these two, and finally how to get around the problem. If there is no way to get around this, NODEs are just highly overrated and far from doing something practical, unless the majority of the mathematical structure of the problem is already known. I should mention that this link will be helpful to the readers encountering the same issue:
[Fitting neural ODE to periodic time series - Specific Domains / Machine Learning - JuliaLang](https://discourse.julialang.org/t/fitting-neural-ode-to-periodic-time-series/42263/3)

I also noticed that ***multiple_shoot*** functionality is not compatible with the ***AugmentedNDELayer*** but I may be wrong. 

In the following, I briefly explained what I tried and what failed:

Since I was not able to use ***AugmentedNDELayer*** with ***multiple_shoot***, I just defined a network like:

```
nn = FastChain((x, p) -> x, #  vcat(x,0.0)
                  FastDense(2, 20, swish),
                  FastDense(20, 40, swish),
                  FastDense(40, 80, swish),
                  FastDense(80, 40, swish),
                  FastDense(40, 20, swish),
                  FastDense(20, 2))
p_init = initial_params(nn)
...
```
and then trained the first dimension on the first dimension of the true data, and left the second dimension out of the loss function. In this case, even if you don't train your network on the second dimension, you have to have it in your training data (even if not using it all in the loss function), since ***multiple_shoot*** needs the second dimension too, to have an initial condition for each group of the second dimension to start training. 

What I observed, if I use the whole training data  produced by ***trueODEfunc***, regardless of using one or two dimensions in the loss function, it very well fits the first dimension at least. 
However, If I use the first dimension of the ***trueODEfunc*** to learn only this dimension, and use another time series as the second dimension (either to learn that too, or just as an augmented space), let it be some sinusoidal function or just a constant zero, no matter how large the network is and no matter if I am using both or just one dimension in the loss faction, the network fails in fitting. 

This is not hard to see why the learning process goes well when both dimensions given by ***trueODEfunc*** as training data. If the two dimensions are used in the loss, the very simple dependency (i.e. true_A = [-0.1 2.0; -2.0 -0.1]) of the two time series is easily found. If only one dimension is used in the loss function, the training process goes well too, since the initial conditions due to grouping process help the network get an approximate trend of the second dimension and this approximation indirectly helps learning the first one. 

But it is not so easy to actually fit the model to the first dimension, when you don't have the second dimension with such a simple dependency to the first. If instead of the second dimension of true data given  by ***trueODEfunc*** we had a simple sinosoidal function, a constant function or a nonlinear function, most likely you're not going to fit any good model to your data. 

Here I bring two versions of the problem, first, a problem that uses all two dimensions given  by ***trueODEfunc*** for training. It fits very well even if the network doesn't use the structure ***u.^3***. The second case, I just replace the second dimension of the data with a sine function or some constant value. It fails to fit miserably!

``` 
using DiffEqFlux, DifferentialEquations, Plots
using DiffEqFlux: group_ranges


# Define initial conditions and time steps
datasize = 100
u0 = Float32[2.0, 0.0]
tspan = (0.0f0, 5.0f0)
tsteps = range(tspan[1], tspan[2], length = datasize)

# Get the data
function trueODEfunc(du, u, p, t)
    true_A = [-0.1 2.0; -2.0 -0.1]
    du .= ((u.^3)'true_A)'
end
prob_trueode = ODEProblem(trueODEfunc, u0, tspan)
ode_data = Array(solve(prob_trueode, Tsit5(), saveat = tsteps))


# Define the Neural Network
nn = FastChain((x, p) -> x, #        vcat(x,0.0)
                  FastDense(2, 20, swish),
                 # FastDense(20, 40, swish),
                 # FastDense(40, 20, swish),
                  FastDense(20, 2))  #  FastDense(20, 1)
p_init = initial_params(nn)



prob_node = ODEProblem((u,p,t)->nn(u,p), u0, tspan, p_init)


function plot_multiple_shoot(plt, preds, group_size)
	step = group_size-1
	ranges = group_ranges(datasize, group_size)

	for (i, rg) in enumerate(ranges)
		plot!(plt, tsteps[rg], preds[i][1,:], markershape=:circle, label="Group $(i)")  #  just changed  this for 1 eq:  from preds[i][1, :] to preds[i][:]
	end
end

# Animate training
anim = Animation()
callback = function (p, l, preds; doplot = true)
  display(l)
  if doplot
	# plot the original data
	plt = scatter(tsteps, ode_data[1,:], legend=nothing)

	# plot the different predictions for individual shoot
	plot_multiple_shoot(plt, preds, group_size)

    frame(anim)
    display(plot(plt))
  end
  return false
end

# Define parameters for Multiple Shooting
group_size = 3
continuity_term = 500


function loss_function(data, pred)
	return sum(abs2, data[1,:] - pred[1,:])  # using only the first dimension in the loss
# return sum(abs2, data- pred)             # using all dimensions 
end

function loss_multiple_shooting(p)
    return multiple_shoot(p, ode_data, tsteps,prob_node, loss_function, Tsit5(),
                          group_size; continuity_term)  #  prob_node
end 


res_ms = DiffEqFlux.sciml_train(loss_multiple_shooting, p_init, cb = callback, maxiters=150)
``` 

A simple change in the data set shows the failure of the training process, just replace the second dimension with a sine wave or a constant function, and make the loss function use one or two dimensions, doesn't help:

``` 

using DiffEqFlux, DifferentialEquations, Plots
using DiffEqFlux: group_ranges


# Define initial conditions and time steps
datasize = 100
u0 = Float32[2.0, 0.0]
tspan = (0.0f0, 5.0f0)
tsteps = range(tspan[1], tspan[2], length = datasize)

# Get the data
function trueODEfunc(du, u, p, t)
    true_A = [-0.1 2.0; -2.0 -0.1]
    du .= ((u.^3)'true_A)'
end
prob_trueode = ODEProblem(trueODEfunc, u0, tspan)
ode_data = Array(solve(prob_trueode, Tsit5(), saveat = tsteps))
ode_data[2,:]  = sin.(2 .+ 5*collect(tsteps)) # just replacing the second dimension with a sine wave or a constant: ode_data[2,:]  .= 1.0
plot(tsteps,[ode_data[1,:],ode_data[2,:]])

# Define the Neural Network
nn = FastChain((x, p) -> x, #        vcat(x,0.0)
                  FastDense(2, 20, swish),
                  FastDense(20, 40, swish),
                  FastDense(40, 20, swish),
                  FastDense(20, 2))  #  FastDense(20, 1)
p_init = initial_params(nn)



prob_node = ODEProblem((u,p,t)->nn(u,p), u0, tspan, p_init)


function plot_multiple_shoot(plt, preds, group_size)
	step = group_size-1
	ranges = group_ranges(datasize, group_size)

	for (i, rg) in enumerate(ranges)
		plot!(plt, tsteps[rg], preds[i][1,:], markershape=:circle, label="Group $(i)")  #  just changed  this for 1 eq:  from preds[i][1, :] to preds[i][:]
	end
end

# Animate training
anim = Animation()
callback = function (p, l, preds; doplot = true)
  display(l)
  if doplot
	# plot the original data
	plt = scatter(tsteps, ode_data[1,:], legend=nothing)

	# plot the different predictions for individual shoot
	plot_multiple_shoot(plt, preds, group_size)

    frame(anim)
    display(plot(plt))
  end
  return false
end

# Define parameters for Multiple Shooting
group_size = 3
continuity_term = 500


function loss_function(data, pred)
	return sum(abs2, data[1,:] - pred[1,:])  # using only the first dimension in the loss
# return sum(abs2, data- pred)             # using all dimensions 
end

function loss_multiple_shooting(p)
    return multiple_shoot(p, ode_data, tsteps,prob_node, loss_function, Tsit5(),
                          group_size; continuity_term)  #  prob_node
end 


res_ms = DiffEqFlux.sciml_train(loss_multiple_shooting, p_init, cb = callback, maxiters=1000)

``` 

Some questions remain: How to fit an NODE to one or more arbitrary time series with no or very complicated dependencies. It turns out that this is not an easy task even in the absence of any stochasticity, sharp changes as in shocks, jumps and stiff problems, or chaotic behaviors. And which method is more effective for learning long time series, multiple shooting? sequential learning? Does the augmenting layer do the same as manually defining one or more dimensions for the network but not training the augmented one(s) in the loss? 
Here we assume that we have little/no knowledge about the mathematical structure and we want an alternative to RC or recurrent network families, we want to make NODEs work!
@ChrisRackauckas

**Comment by ChrisRackauckas** [comment]
Well if you want the real answer here, the continuous-time echo state network training is just way more robust so that's generally the better option. Reservoir computing techniques are just dead simple to "train".

> I also noticed that multiple_shoot functionality is not compatible with the AugmentedNDELayer but I may be wrong.

Make a second issue on this.

> A simple change in the data set shows the failure of the training process, just replace the second dimension with a sine wave or a constant function, and make the loss function use one or two dimensions, doesn't help:

This is precisely the issue discussed in the augmented neural ODE paper. Not every possible timeseries has a valid (or simple) ODE. Specifically, if you look at the phase space diagram and you ever have a line intersect, then that has no valid ODE solution! So the solution is to learn a higher dimensional ODE system so that the first two dimensions match your time series, this is the augmented neural ODE. You only tried to learn a 2-dimensional system to match that, which as discussed is provably not possible in general.

> Does the augmenting layer do the same as manually defining one or more dimensions for the network but not training the augmented one(s) in the loss?

I think this is your misunderstanding here. It's not adding another dimension to the network, it's another ODE equation. You need to raise the dimension of the ODE itself.

**Comment by ashaygan** [comment]
@ChrisRackauckas Thanks a lot Chris, your comments, always helpful and comprehensive. For sure, I'll open another issue for the augmented layer compatibility with the multiple shooting method and close this. Sorry for my misunderstanding for that part, you are absolutely right.
